### PR TITLE
cv sub forms

### PR DESCRIPTION
**Pull request by marijana-m**
added cv_field_array.tsx as sub form (for certifications and personal_qualities)
added projects.tsx sub form
added technical_skill.tsx for more flexibility
instead of separate arrays for programming_languages, libs_and_frameworks, big_data, databases and devops added a new technical_skills array that contains skills objects that user can input
user can add a skill group, change skill group name and add skills to that group

Screen Shot 2023-01-26 at 4 08 19 PM

new technical_skills table: https://app.supabase.com/project/lcfztnwcvbwktlagmxdm/editor/26145

### CODE REVIEW

File: components/cv-form/technical_skill.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                  <div className="mb-6 flex flex-shrink-0 flex-col md:mb-0 md:w-64">
+                    <span className="title-font font-semibold text-gray-700 dark:text-gray-400">
+                      Skills
+                    </span>
+                  </div>
+                  <div className="md:flex-grow">
+                    <FieldArray
+                      name={`technical_skills[${index}].skills`}
+                      render={(arrayHelpers) => (
+                        <div>
```
Comment (akoskm) [inline components/cv-form/technical_skill.tsx:46]: Why do we need the wrapping div?

File: components/cv-form/technical_skill.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                                  <button
+                                    className="rounded-md border border-indigo-500 bg-indigo-500 p-1  text-white"
+                                    type="button"
+                                    onClick={() =>
+                                      arrayHelpers.remove(skillIndex)
+                                    }
+                                  >
+                                    <Icons.TrashCan />
+                                  </button>
+                                </div>
```
Comment (akoskm) [inline components/cv-form/technical_skill.tsx:69]: This should be a separate `SkillRow` component.

File: components/cv-form/projects.tsx [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+                                  <button
+                                    className="rounded-md border border-indigo-500 bg-indigo-500 p-1  text-white"
+                                    type="button"
+                                    onClick={() =>
+                                      arrayHelpers.remove(responsibilitiesIndex)
+                                    }
+                                  >
+                                    <Icons.TrashCan />
+                                  </button>
+                                </div>
```
Comment (akoskm) [inline components/cv-form/projects.tsx:189]: Responsibility row component.

File: components/cv-form/technical_skill.tsx (hunk 1 above)
Comment (marijana-m) [reply components/cv-form/technical_skill.tsx:46]: to wrap the button below for adding skill, I used a div but I can also use <></>, just some wrapper was necessary

File: components/cv-form/technical_skill.tsx (hunk 2 above)
Comment (marijana-m) [reply components/cv-form/technical_skill.tsx]: added nested_row reusable component for skills array

File: components/cv-form/projects.tsx (hunk 3 above)
Comment (marijana-m) [reply components/cv-form/projects.tsx]: added nested_row reusable component for responsibility array, and also for technologies array

File: components/add_new_cv_form.tsx [hunk 4]
```diff
@@ -4,10 +4,12 @@ import supabase from "../utils/supabase_browser";
 import { useEffect, useState } from "react";
 import { useRouter } from "next/navigation";
 import { CV } from "./types";
-import * as Icons from "./Icons";
 import * as Options from "../constants/CvFormOptions";
-import { Formik, Field, FieldArray, Form, ErrorMessage } from "formik";
+import { Formik, Field, Form, ErrorMessage } from "formik";
 import * as Yup from "yup";
+import CvFieldArray from "./cv-form/cv_field_array";
```
Comment (akoskm) [inline components/add_new_cv_form.tsx:10]: Name the folder `cv_form` to comply with the file naming pattern.

File: components/add_new_cv_form.tsx [hunk 5]
```diff
@@ -108,9 +98,22 @@ export default function AddNewCvForm({ id }: Props) {
...
+    }
+    return supabase
+      .from("technical_skills")
+      .insert({ ...values, cv_id: cvData[0].id });
+  }
+
   async function handleSubmit(values: any) {
     const { data, error } = await upsert(values);
     values.projects.forEach((project: any) => upsertProjects(project, data));
+    values.technical_skills.forEach((skill: any) => upsertSkills(skill, data));
```
[the remark this answers is not in the document]
Comment (marijana-m) [reply components/add_new_cv_form.tsx:116]: I will add error handling for skills

File: components/add_new_cv_form.tsx [hunk 6]
```diff
@@ -97,26 +86,33 @@ export default function AddNewCvForm({ id }: Props) {
...
 
   async function handleSubmit(values: any) {
     const { data, error } = await upsert(values);
-    values.projects.forEach((project: any) => upsertProjects(project, data));
-
-    if (error) {
-      setServerErrorMessage(error.message);
-    } else {
-      await router.push("/");
-    }
```
Comment (akoskm) [inline components/add_new_cv_form.tsx:119]: Why did you remove the error handling for projects?

Comment (marijana-m) [reply components/add_new_cv_form.tsx:119]: because I already changed projects structure so upsertProject function here doesn't do anything, I commented that line now
i added server errors for projects in [this PR](https://github.com/bditengineering/cv-app/pull/16) with an updated upsertProjects function

Comment (akoskm) [reply components/add_new_cv_form.tsx:119]: Understood. Why did you comment on the line instead of deleting it?

File: components/add_new_cv_form.tsx [hunk 7]
```diff
@@ -97,26 +86,33 @@ export default function AddNewCvForm({ id }: Props) {
...
     return supabase.from("cv").upsert(updatedCv).select();
   }
 
   async function upsertProjects(values: any, cvData: any) {
-    if (values.id) {
-      return supabase.from("projects").update(values).eq("id", values.id);
-    }
-    return supabase.from("projects").insert({ ...values, cv_id: cvData[0].id });
+    values.cv_id = cvData[0].id;
+    return supabase.from("projects").upsert(values);
```
Comment (akoskm) [inline components/add_new_cv_form.tsx:96]: So this function is no longer used in this PR per your comment below.

Can we remove it?

File: components/add_new_cv_form.tsx (hunk 6 above)
Comment (marijana-m) [reply components/add_new_cv_form.tsx:119]: no particular reason, deleted that line now instead

File: components/add_new_cv_form.tsx (hunk 7 above)
Comment (marijana-m) [reply components/add_new_cv_form.tsx]: sure, deleted it now

**Review by akoskm — APPROVED** [review]
